File: 0823. Binary Trees With Factors/Solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int numFactoredBinaryTrees(int[] arr) {
        int P = 1000000007;
        Arrays.sort(arr);

        Map<Integer, Long> counter = new HashMap<>();
        for (int e : arr) {
            counter.put(e, 1L);
        }
        for (int i = 1; i < arr.length; i++) {
            for (int j = 0; j < i; ++j) {
                if (arr[i] % arr[j] == 0) {
                    int a = arr[j], b = arr[i] / arr[j];
                    if (counter.containsKey(b)) {
                        counter.put(arr[i], (counter.get(arr[i]) + counter.get(a) * counter.get(b)) % P);
                    }
                }
            }
        }

        long res = 0L;
        for (Long e : counter.values()) {
            res = (res + e) % P;
        }
        return (int) res;
    }
}

File: 0823. Binary Trees With Factors/Solution.cpp
#include <algorithm>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr)
    {
        int P = 1000000007;
        unordered_map<int, long long> cnt;
        for (const int& e : arr)
            cnt[e] = 1;
        sort(arr.begin(), arr.end());
        for (size_t i = 1; i < arr.size(); ++i) {
            for (size_t j = 0; j < i; ++j) {
                if (arr[i] % arr[j] != 0)
                    continue;
                int a = arr[j], b = arr[i] / arr[j];
                if (!cnt.count(b))
                    continue;
                cnt[arr[i]] = (cnt[arr[i]] + cnt[a] * cnt[b]) % P;
            }
        }

        int res = 0;
        for (const auto& it : cnt)
            res = (res + it.second) % P;
        return res;
    }
};
